### PR TITLE
shared/sleep-config,hibernate-util: cleanup round two

### PR DESCRIPTION
**Pull request by YHNdnzj**

### CODE REVIEW

File: src/shared/hibernate-util.c [hunk 1]
```diff
@@ -28,446 +28,425 @@
...
+        assert(hibernate_device);
 
-        return mfree(se);
+        free(hibernate_device->path);
 }
 
-HibernateLocation* hibernate_location_free(HibernateLocation *hl) {
-        if (!hl)
-                return NULL;
+int read_fiemap(int fd, struct fiemap **ret) {
```
Comment (bluca) [inline src/shared/hibernate-util.c:38]: read_filemap?

**Review by bluca — APPROVED** [review]
Also rebase on main so that we get a testing run on jammy-amd64

**Comment by bluca** [comment]
'dontmerge' label is set, is that intended?

**Comment by YHNdnzj** [comment]
> 'dontmerge' label is set, is that intended?

Yes. This is not the final version yet. Should have removed the `please-review` label...

**Comment by bluca** [comment]
You can convert it to 'draft' so that the bot doesn't add the label

File: src/shared/hibernate-util.c (hunk 1 above)
Comment (YHNdnzj) [reply src/shared/hibernate-util.c:38]: Hmm, I didn't touch this part. And I think it makes sense to say `fiemap` as the struct name.

File: src/shared/sleep-config.h [hunk 2]
```diff
@@ -35,7 +35,19 @@ DEFINE_TRIVIAL_CLEANUP_FUNC(SleepConfig*, sleep_config_free);
 
 int parse_sleep_config(SleepConfig **sleep_config);
 
-int can_sleep(SleepOperation operation);
+typedef enum SleepSupport {
+        SLEEP_OK,
+        SLEEP_DISABLED,
+        SLEEP_NOTCONFIGURED,
```
Comment (poettering) [inline src/shared/sleep-config.h]: space between NOT and CONFIGURED?

Comment (poettering) [reply src/shared/sleep-config.h]: can you add a comment here what the difference between these two is?

File: src/shared/sleep-config.h [hunk 3]
```diff
@@ -35,7 +35,19 @@ DEFINE_TRIVIAL_CLEANUP_FUNC(SleepConfig*, sleep_config_free);
...
+        SLEEP_POWER_UNSUPPORTED,
+        SLEEP_NO_ENOUGH_SWAP_SPACE,
```
Comment (poettering) [inline src/shared/sleep-config.h]: NO → NOT

File: src/shared/sleep-config.h [hunk 4]
```diff
@@ -35,7 +35,19 @@ DEFINE_TRIVIAL_CLEANUP_FUNC(SleepConfig*, sleep_config_free);
 
 int parse_sleep_config(SleepConfig **sleep_config);
 
-int can_sleep(SleepOperation operation);
+typedef enum SleepSupport {
+        SLEEP_OK,
```
Comment (poettering) [inline src/shared/sleep-config.h]: shouldn't this be YES?

File: src/login/logind-dbus.c [hunk 5]
```diff
@@ -1923,16 +1923,34 @@ static int method_do_shutdown_or_sleep(
...
+                if (r == 0)
+                        switch (support) {
+
+                        case SLEEP_DISABLED:
+                                return sd_bus_error_setf(error, BUS_ERROR_SLEEP_VERB_NOT_SUPPORTED,
+                                                         "Sleep verb '%s' is disabled by config",
+                                                         sleep_operation_to_string(a->sleep_operation));
+
+                        case SLEEP_NOTCONFIGURED:
+                        case SLEEP_POWER_UNSUPPORTED:
```
Comment (poettering) [inline src/login/logind-dbus.c]: hmm, "POWER"? maybe call this "OPERATION"?

File: src/shared/sleep-config.c [hunk 6]
```diff
@@ -110,82 +112,81 @@ int parse_sleep_config(SleepConfig **ret) {
...
-                log_debug("Sleep mode %s not supported by the kernel, sorry.", strnull(t));
+        if (r > 0) {
+                log_debug("Sleep state '%s' is supported by kernel.", found);
+                return true;
         }
-        return r;
+
+        if (DEBUG_LOGGING) {
+                _cleanup_free_ char *joined = strv_join(states, " ");
+                log_debug("Configured sleep states are unsupported by kernel: %s", strnull(joined));
```
Comment (keszybz) [inline src/shared/sleep-config.c]: "None of the configured sleep states is supported by the kernel"

("unsupported" doesn't sound good. "not supported" would be better, but I think it's better to reword this to clarify that it's an OR condition.)

File: src/shared/sleep-config.c [hunk 7]
```diff
@@ -110,82 +112,81 @@ int parse_sleep_config(SleepConfig **ret) {
...
+                        log_debug("Disk sleep mode '%s' is supported by kernel.", mode);
                         return true;
                 }
         }
 
         if (DEBUG_LOGGING) {
-                _cleanup_free_ char *t = strv_join(types, "/");
-                log_debug("Disk sleep mode %s not supported by the kernel, sorry.", strnull(t));
+                _cleanup_free_ char *joined = strv_join(modes, " ");
+                log_debug("Configured disk sleep modes are unsupported by kernel: %s", strnull(joined));
```
Comment (keszybz) [inline src/shared/sleep-config.c]: Also here, "unsupported" is not an improvement over the previous text.

File: src/shared/sleep-config.h [hunk 8]
```diff
@@ -35,7 +35,19 @@ DEFINE_TRIVIAL_CLEANUP_FUNC(SleepConfig*, sleep_config_free);
...
 int parse_sleep_config(SleepConfig **sleep_config);
 
-int can_sleep(SleepOperation operation);
+typedef enum SleepSupport {
+        SLEEP_SUPPORTED,
+        SLEEP_DISABLED, /* Disabled in SleepConfig.allow */
+        SLEEP_NOT_CONFIGURED, /* SleepConfig.states is not configured */
+        SLEEP_STATE_OR_MODE_UNSUPPORTED, /* SleepConfig.states/modes are not supported by kernel */
+        SLEEP_NOT_ENOUGH_SWAP_SPACE,
+        SLEEP_ALARM_UNSUPPORTED, /* CLOCK_BOOTTIME_ALARM is unsupported by kernel (only used by s2h) */
```
Comment (keszybz) [inline src/shared/sleep-config.h]: Also here, "not supported".

Maybe align the comments to the same column?

File: src/shared/hibernate-util.c [hunk 9]
```diff
@@ -28,446 +28,414 @@
...
+                           "\n",
+                           &swap.path, &type, &swap.size, &swap.used, &swap.priority);
                 if (k == EOF)
                         break;
-                if (k != 5) {
-                        log_debug("Failed to parse /proc/swaps:%u, ignoring", i);
-                        continue;
-                }
+                if (k != 5)
+                        return log_debug_errno(SYNTHETIC_ERRNO(EIO), "Failed to parse /proc/swaps: %m");
```
Comment (keszybz) [inline src/shared/hibernate-util.c]: %m shouldn't be used here. Also, the previous message printed the offending line, which was quite useful.

File: src/shared/hibernate-util.c [hunk 10]
```diff
@@ -28,446 +28,414 @@
...
                 } else if (streq(type, "partition")) {
-                        const char *fn;
+                        const char *node;
 
-                        fn = path_startswith(swap->path, "/dev/");
-                        if (fn && startswith(fn, "zram")) {
-                                log_debug("%s: ignoring zram swap", swap->path);
+                        node = path_startswith(swap.path, "/dev/");
+                        if (node && startswith(node, "zram")) {
+                                log_debug("Swap partition '%s' is zram device, ignoring.", swap.path);
```
Comment (keszybz) [inline src/shared/hibernate-util.c]: "is a zram device"

File: src/shared/hibernate-util.c [hunk 11]
```diff
@@ -28,446 +28,414 @@
...
 
-        /* We found nothing at all */
-        if (!hibernate_location)
-                return log_debug_errno(SYNTHETIC_ERRNO(ENOSYS),
-                                       "No possible swap partitions or files suitable for hibernation were found in /proc/swaps.");
+        r = read_swap_entries(&entries);
+        if (r < 0)
+                return r;
+        if (entries.n_swaps == 0)
+                return log_debug_errno(SYNTHETIC_ERRNO(ENOENT), "No swap space available for hibernation found.");
```
Comment (keszybz) [inline src/shared/hibernate-util.c]: Drop "found".

File: src/shared/hibernate-util.c [hunk 12]
```diff
@@ -28,446 +28,414 @@
...
                 log_debug_errno(r, "Failed to retrieve Active(anon) from /proc/meminfo: %m");
                 return false;
         }
 
-        r = safe_atollu(active, &act);
+        r = safe_atollu(active_str, &active);
         if (r < 0) {
-                log_debug_errno(r, "Failed to parse Active(anon) from /proc/meminfo: %s: %m", active);
+                log_debug_errno(r,
+                                "Failed to parse Active(anon) '%s' from /proc/meminfo, assuming no enough space: %m",
```
Comment (keszybz) [inline src/shared/hibernate-util.c]: "not enough"

File: src/login/logind-dbus.c [hunk 13]
```diff
@@ -1943,6 +1943,10 @@ static int method_do_shutdown_or_sleep(
                                                          "Sleep verb '%s' is not configured or configuration is unsupported by kernel",
                                                          sleep_operation_to_string(a->sleep_operation));
 
+                        case SLEEP_RESUME_UNSUPPORTED:
```
Comment (keszybz) [inline src/login/logind-dbus.c]: "NOT_SUPPORTED"

File: src/login/logind-dbus.c [hunk 14]
```diff
@@ -1943,6 +1943,10 @@ static int method_do_shutdown_or_sleep(
...
+                                return sd_bus_error_setf(error, BUS_ERROR_SLEEP_VERB_NOT_SUPPORTED,
+                                                         "Not EFI booted and resume= is not set. No available method to resume from hibernation");
```
Comment (keszybz) [inline src/login/logind-dbus.c]: "EFI booted" is very jargony. Maybe just say "Not running on EFI"

File: src/shared/hibernate-util.c [hunk 15]
```diff
@@ -402,38 +403,53 @@ int find_suitable_hibernate_device_full(HibernateDevice *ret_device, uint64_t *r
...
+        resume_set = r > 0;
 
-        r = get_proc_field("/proc/meminfo", "Active(anon)", WHITESPACE, &active_str);
-        if (r < 0) {
-                log_debug_errno(r, "Failed to retrieve Active(anon) from /proc/meminfo: %m");
-                return false;
+        if (!resume_set && !is_efi_boot()) {
+                log_debug("Not EFI booted and resume= is not set. Hibernation is not safe.");
+                return -ENOTRECOVERABLE;
         }
```
Comment (keszybz) [inline src/shared/hibernate-util.c]: "Not running on EFI…"

`return log_debug_errno(SYNTHETIC_ERRNO(…))`

File: src/shared/hibernate-util.h [hunk 16]
```diff
@@ -18,7 +18,7 @@ static inline int find_suitable_hibernate_device(HibernateDevice *ret) {
         return find_suitable_hibernate_device_full(ASSERT_PTR(ret), NULL, NULL);
 }
 
-bool enough_swap_for_hibernation(void);
+int hibernate_is_safe(void);
```
Comment (keszybz) [inline src/shared/hibernate-util.h]: "hibernate" is a verb. Please call this "hibernation_is_safe".

**Review by keszybz — CHANGES_REQUESTED** [review]
I was confused by some of the messages getting rewritten multiple times…

File: src/shared/sleep-config.c [hunk 17]
```diff
@@ -110,82 +112,81 @@ int parse_sleep_config(SleepConfig **ret) {
...
-                log_debug("Sleep mode %s not supported by the kernel, sorry.", strnull(t));
+        if (r > 0) {
+                log_debug("Sleep state '%s' is supported by kernel.", found);
+                return true;
         }
-        return r;
+
+        if (DEBUG_LOGGING) {
+                _cleanup_free_ char *joined = strv_join(states, " ");
+                log_debug("None of the configured sleep states is supported by kernel: %s", strnull(joined));
```
Comment (keszybz) [inline src/shared/sleep-config.c]: Oops, I think it's my bad suggestion, but "None … are".

File: src/shared/sleep-config.c [hunk 18]
```diff
@@ -110,82 +112,81 @@ int parse_sleep_config(SleepConfig **ret) {
...
+                        log_debug("Disk sleep mode '%s' is supported by kernel.", mode);
                         return true;
                 }
         }
 
         if (DEBUG_LOGGING) {
-                _cleanup_free_ char *t = strv_join(types, "/");
-                log_debug("Disk sleep mode %s not supported by the kernel, sorry.", strnull(t));
+                _cleanup_free_ char *joined = strv_join(modes, " ");
+                log_debug("None of the configured hibernation power modes is supported by kernel: %s", strnull(joined));
```
Comment (keszybz) [inline src/shared/sleep-config.c]: Same here.

File: src/shared/hibernate-util.c [hunk 19]
```diff
@@ -28,446 +28,412 @@
...
-                }
+                if (k != 5)
+                        return log_debug_errno(SYNTHETIC_ERRNO(EIO), "Failed to parse /proc/swaps line %u.", i);
 
                 if (streq(type, "file")) {
-
-                        if (endswith(swap->path, "\\040(deleted)")) {
-                                log_debug("Ignoring deleted swap file '%s'.", swap->path);
+                        if (endswith(swap.path, "\\040(deleted)")) {
+                                log_debug("Swap file '%s' is deleted, ignoring.", swap.path);
```
Comment (keszybz) [inline src/shared/hibernate-util.c]: "is" → "has been" ?

File: src/shared/hibernate-util.h [hunk 20]
```diff
@@ -4,39 +4,23 @@
...
-        SWAP_BLOCK,
-        SWAP_FILE,
-        _SWAP_TYPE_MAX,
-        _SWAP_TYPE_INVALID = -EINVAL,
-} SwapType;
-
-/* entry in /proc/swaps */
-typedef struct SwapEntry {
+/* represents values for /sys/power/resume & /sys/power/resume_offset and the corresponding path */
+typedef struct HibernateDevice {
```
Comment (keszybz) [inline src/shared/hibernate-util.h]: "HibernationDevice"

**Review by keszybz — CHANGES_REQUESTED** [review]
LGTM, just some nitpicks.

**Review by bluca — APPROVED** [review]

File: src/shared/hibernate-util.c [hunk 21]
```diff
@@ -28,441 +29,426 @@
...
-                return errno == ENOENT ? -EOPNOTSUPP : -errno; /* Convert swap not supported to a recognizable error */
+        r = get_block_device_fd(fd, &swap->devno);
+        if (r < 0)
+                return r;
+
+        r = fd_is_fs_type(fd, BTRFS_SUPER_MAGIC);
+        if (r < 0)
+                return log_debug_errno(r, "Failed to check if swap file '%s' is on Btrfs: %m", swap->path);
+        if (r > 0) {
+                r = btrfs_get_file_physical_offset_fd(fd, &offset_raw);
```
Comment (MartinNowak) [inline src/shared/hibernate-util.c:234]: The usage of btrfs_get_file_physical_offset_fd when checking the configured kernel resume offset now requires CAP_SYS_ADMIN for systemd-sleep hibernate @YHNdnzj.

```
allow systemd_sleep_t self:capability sys_admin;
```
Fedora's SELinux policy only allows systemd_sleep_t to use the following capabilities.
```
allow systemd_sleep_t self:capability { linux_immutable sys_resource };
# systemd-sleep needs to set timer for suspend-then-hibernate
allow systemd_sleep_t self:capability2 wake_alarm;
dontaudit systemd_sleep_t self:capability sys_ptrace;
```
https://github.com/fedora-selinux/selinux-policy/blob/85bbb6d1bc69e84ea2440cff647e426b74f1b40a/policy/modules/system/systemd.te#L1594-L1597

The first ioctl that fails is [BTRFS_IOC_TREE_SEARCH](https://btrfs.readthedocs.io/en/latest/btrfs-ioctl.html).

**Review by MartinNowak — COMMENTED** [review]
It seems that this change broke hibernation on btrfs in Fedora 40's beta.

File: src/shared/hibernate-util.c (hunk 21 above)
Comment (YHNdnzj) [reply src/shared/hibernate-util.c:234]: Hmm, sorry, but I have basically 0 knowledge about SELinux... Could you please open an issue there? I don't think there's anything to change in systemd codebase.

Comment (daandemeyer) [reply src/shared/hibernate-util.c:234]: Yeah this should be fixed in Fedora's selinux policy.

Comment (MartinNowak) [reply src/shared/hibernate-util.c:234]: I merely wonder if it was possible without CAP_SYS_ADMIN or should fallback to skip the check when lacking permission. I do see, that a lot of other systemd services have sys_admin capability, so will make a PR to fix the SELinux policy.

Comment (YHNdnzj) [reply src/shared/hibernate-util.c:234]: > I merely wonder if it was possible without CAP_SYS_ADMIN or should fallback to skip the check when lacking permission. 

Well, this is more about auto-detecting resume offset than checking the existing value. So I don't think it's desirable.

> I do see, that a lot of other systemd services have sys_admin capability, so will make a PR to fix the SELinux policy.

Thanks!
